### PR TITLE
Fix the uncontrolled blitting which defeats the copy contructor

### PR DESCRIPTION
**Pull request by SSoulaimane**
Issue [20321](https://issues.dlang.org/show_bug.cgi?id=20321).

LDC and GDC are also affected.

I chose this way of fixing it which keeps the initialization at the semantic phase then just adapts it in the codegen, but I realized that the backend creates a lot of temporaries and copies into them. There has to be some means to do elaborate copy in the backend, I'm afraid there is no workaround.

CC: @WalterBright 
CC: @kinke @ibuclaw are you going to solve this for LDC and GDC? How?

### CODE REVIEW

**Comment by ibuclaw** [comment]
What's the frontend representational dump of this?  By and large the code gen just does what it is told, and only creates temporaries when a side effect needs to be evaluated multiple times.

**Comment by SSoulaimane** [comment]
Here is an example which illustrates the needs call the copy constructor in the codegen.
compiling this snippet
```d
Container[] darr;
darr ~= v;
```
hits the following lines
https://github.com/dlang/dmd/blob/37a0434cfb86899285f7c633bddf0c2b702563cd/src/dmd/e2ir.d#L3315-L3326
The codegen evaluates `v` into a temporary first then moves it into the array. It should call the copy constructor or ideally the move constructor when it moves from the temporary into the array.

**Comment by SSoulaimane** [comment]
@ibuclaw That's true, the copying is the problem, it's doesn't call the copy constructor when it does it.
I hand-picked those test cases by following usage of `doCopyOrMove()` you can start with the low hanging fruit which is struct function arguments for Posix. For Win32 calling conventions you need more backend intervention to get it right but on the bright side GCC and Clang already do it correctly with C++.

File: src/dmd/backend/cdef.d [hunk 1]
```diff
@@ -875,7 +875,11 @@ union eve
         {
             elem* E1;           // left child for unary & binary nodes
             elem* E2;           // right child for binary nodes
-            Symbol* Edtor;      // OPctor: destructor
+            union
+            {
+                Symbol* Edtor;  // OPctor: destructor
+                Symbol* Edecl2; // VarDeclaration being constructed
```
Comment (WalterBright) [inline src/dmd/backend/cdef.d]: When is Edecl2 valid?

**Comment by WalterBright** [comment]
> Issue 20321

Needs to be in the comment message and the subject, or the bot won't work.

File: test/runnable/internalpointer.d [hunk 2]
```diff
@@ -0,0 +1,59 @@
+/** Container with internal pointer
```
Comment (WalterBright) [inline test/runnable/internalpointer.d]: Give url to bugzilla this is fixing.

File: src/dmd/e2ir.d [hunk 3]
```diff
@@ -6362,6 +6454,65 @@ elem* elAssign(elem* e1, elem* e2, Type t, type* tx)
...
+ *    `false` otherwise.
+ */
+private bool assignOPstrctor(ref elem* e1, ref elem* e2)
+{
+    size_t n;
+    void recurse(elem* e1, ref elem* e2)
+    {
+        if (e2.Eoper == OPstrctor)
+        {
+            /* Elaborate copy or assignment for aggregate rvalues.
```
Comment (WalterBright) [inline src/dmd/e2ir.d]: What is "elaborate" ?

File: src/dmd/e2ir.d [hunk 4]
```diff
@@ -6362,6 +6454,65 @@ elem* elAssign(elem* e1, elem* e2, Type t, type* tx)
     return e;
 }
 
+
+/**********************************
+ * Assign OPstrctor in `e2` it to `e1`.
```
Comment (WalterBright) [inline src/dmd/e2ir.d]: This sentence makes no sense.

Comment (WalterBright) [reply src/dmd/e2ir.d]: I have no idea what this function does.

File: src/dmd/e2ir.d [hunk 5]
```diff
@@ -5932,6 +5977,52 @@ private elem *appendDtors(IRState *irs, elem *er, size_t starti, size_t endi)
...
+    void findReplace(ref elem* e)
+    {
+        if (OTleaf(e.Eoper))
+        {
+            // use `s` to identify the related elems
+            if (e.Eoper == OPstrthis && e.EV.Vsym == s)
+            {
+                e = rep;
+                if (n == 0)
+                    rep = el_same(&e);
```
Comment (WalterBright) [inline src/dmd/e2ir.d]: this is weird - why is `rep` being rewritten?

File: src/dmd/e2ir.d [hunk 6]
```diff
@@ -5932,6 +5977,52 @@ private elem *appendDtors(IRState *irs, elem *er, size_t starti, size_t endi)
...
+                else
+                    rep = el_copytree(rep);
+                ++n;
+            }
+        }
+        else
+        {
+            findReplace(e.EV.E1);
+            if (OTbinary(e.Eoper))
+                findReplace(e.EV.E2);
```
Comment (WalterBright) [inline src/dmd/e2ir.d]: Use a loop instead of tail recursion to keep stack usage down.

File: src/dmd/e2ir.d [hunk 7]
```diff
@@ -5932,6 +5977,52 @@ private elem *appendDtors(IRState *irs, elem *er, size_t starti, size_t endi)
...
+/*************************************
+ * Find all occurrences of OPstrthis in elem tree `e`,
+ * and replace each with `rep`.
+ *
+ * Params:
+ *     e = OPstrctor elem.
+ *     rep = replacement for OPstrthis.
+ *
+ * Returns:
+ *     number of replacement performed.
```
Comment (WalterBright) [inline src/dmd/e2ir.d]: replacements

File: src/dmd/e2ir.d [hunk 8]
```diff
@@ -5307,6 +5339,19 @@ elem *toElem(Expression e, IRState *irs)
                 s = s.toAlias();
                 if (s != vd)
                     return Dsymbol_toElem(s);
+                if (vd.ismemproxy)
+                {
+                    /* Generate OPstrctor
+                     */
+                    Symbol *sp = toSymbol(s);
+                    assert(vd._init && vd._init.isExpInitializer());
+                    auto ie = vd._init.isExpInitializer();
```
Comment (WalterBright) [inline src/dmd/e2ir.d]: don't need to call isExpInitializer() twice

File: src/dmd/e2ir.d [hunk 9]
```diff
@@ -1175,6 +1177,27 @@ elem *toElem(Expression e, IRState *irs)
                 nrvo = 1;
             }
 
+            /* Placeholder to receive the address of an lvalue
```
Comment (WalterBright) [inline src/dmd/e2ir.d]: ???

File: src/dmd/backend/cod1.d [hunk 10]
```diff
@@ -4206,6 +4214,27 @@ void pushParams(ref CodeBuilder cdb, elem* e, uint stackalign, tym_t tyf)
         }
     }
 
+    version (MARS)
+    {
+        case OPstrctor:
+        {
```
Comment (WalterBright) [inline src/dmd/backend/cod1.d]: What is this block of code doing?

**Comment by WalterBright** [comment]
Note this is quite a bit more complex than the ldc solution: https://github.com/ldc-developers/ldc/pull/3204

Which one is better?

**Comment by ibuclaw** [comment]
Some observations:

1. gdc-9 doesn't have the new cpctor feature, so of course things will work differently.  As the frontend doesn't report the struct as non-POD, it will be copied around in the backend.

2. It's the frontend that's adding in the temporaries, this is a diff between gdc-9 and gdc-master.

```
 Container get()
 {
 	Container a = a = 0 , a.this(1);
-	Container b = a;
+	Container b = b = 0 , b.this(a);
 	a.check();
 	b.check();
 	if (1)
-		return a;
+		return (Container __copytmp2 = __copytmp2 = 0 , __copytmp2.this(a);) , __copytmp2;
 	else
-		return b;
+		return (Container __copytmp3 = __copytmp3 = 0 , __copytmp3.this(b);) , __copytmp3;
 }
 void main()
 {
 	Container v = v = 0 , v.this(1);
 	v.check();
-	func(v);
+	func(((Container __copytmp5 = __copytmp5 = 0 , __copytmp5.this(v);) , __copytmp5));
 	Container r = get();
 	r.check();
-	Container[1] slit = [v];
+	Container[1] slit = [((Container __copytmp6 = __copytmp6 = 0 , __copytmp6.this(v);) , __copytmp6)];
 	slit[0].check();
-	Container[] dlit = [v];
+	Container[] dlit = [((Container __copytmp7 = __copytmp7 = 0 , __copytmp7.this(v);) , __copytmp7)];
 	dlit[0].check();
-	B b = B(v);
+	B b = B(((Container __copytmp8 = __copytmp8 = 0 , __copytmp8.this(v);) , __copytmp8));
 	b.m.check();
 	return 0;
 }
```

3. In gdc-master, the frontend knows about, and reports cpctors as making a struct non-POD, and so these are not copied when passed around directly.  The error on line 20 therefore doesn't happen.

**Comment by ibuclaw** [comment]
Regarding the other failures (from the bug report)

```
test.d(53): error
test.d(56): error
```
`get()` fails because the frontend insists on returning a __copytmp.  I think it would be much better if the frontend recognized that any struct with a cpctor returned is _always_ nrvo, even if there are multiple possible variables.

```
test.d(59): error
test.d(62): error
```
Arrays fail because again of the __copytmps, it would be better if the frontend recognized and passed the address of all indexes to the cpctor instead.

```
test.d(65): error
```
Structs with non-POD members are again failing because of __copytmp, frontend should recognize and pass address of each relevant field to the cpctor.

**Comment by kinke** [comment]
> are you going to solve this for LDC and GDC? How?

It can be solved for LDC, pretty easily as Walter noted, by
1. changing the extern(D) ABI to pass non-POD value arguments by reference under the hood (just like the C++ ABI, except on Win32), and
2. by detecting constructions from temporaries (`auto var = (T __copytmp = __copytmp = … , … , __copytmp)`) and replacing the temporary's lvalue/alloca by the final lvalue, after `toElem()`enting everything as usual; that's a neat feature of LLVM's IR.

File: src/dmd/backend/cod1.d (hunk 10 above)
Comment (SSoulaimane) [reply src/dmd/backend/cod1.d]: The same thing it does for DMC, see line 4172.

File: src/dmd/e2ir.d (hunk 4 above)
Comment (SSoulaimane) [reply src/dmd/e2ir.d]: I moved it to `cgelem.d` where it makes more sense. Check it now, it's a tree optimization.

File: src/dmd/backend/cdef.d (hunk 1 above)
Comment (SSoulaimane) [reply src/dmd/backend/cdef.d]: Edecl2 holds the temporary symbol being constructed. It's just used to find which `this` nodes correspond to which `ctor` node without assuming much about the tree. Nodes which share the same Edecl2 symbol are related together.

File: src/dmd/e2ir.d (hunk 9 above)
Comment (SSoulaimane) [reply src/dmd/e2ir.d]: hopefully it's clearer now.

File: src/dmd/backend/cod1.d (hunk 10 above)
Comment (SSoulaimane) [reply src/dmd/backend/cod1.d]: This does the work for the Win32 ABI. It fills `OPstrthis` nodes with the offset of the stack arguments, we need this to be able to call the copy contractor directly on the area here arguments are pushed. DMC does this with non-POD C++ arguments and it works.

**Comment by SSoulaimane** [comment]
@WalterBright LDC doesn't solve Win32 ABI yet, this solves if for DMD. For the rest it's similar in concept to how LDC does it, LDC has this function https://github.com/ldc-developers/ldc/blob/baaef17472bdeae90f80fb5c4f3b0212971db7a5/gen/toir.cpp#L2767 which is called a number of times. This is a DMD variant using `OPstrctor`. Since we don't have in the backend a handy container for holding and representing values that is used ubiquitously like LDC uses `DLValue`, then trying to mimic LDC here will involve building something like `DLvalue` and spreading it across the backend.

**Comment by SSoulaimane** [comment]
@ibuclaw if you don't have the copy constructor try with the postblit it's the same.

**Comment by SSoulaimane** [comment]
@ibuclaw In summary you're proposing to move much of the work done in the glue layer into the semantic phase. I'm not sure that would end the story though unless a no copying rule is imposed on the glue and the codegen.

File: src/dmd/backend/cgelem.d [hunk 11]
```diff
@@ -3821,6 +3828,177 @@ private elem * elnegass(elem *e, goal_t goal)
...
+        }
+    }
+}
+
+/***************************
+ * Walk tree, find OPstrctor and replace 'this'
+ * argument with `e1`.
+ */
+
+void elstrctorwalk(elem *e, elem* e1)
```
Comment (WalterBright) [inline src/dmd/backend/cgelem.d]: private?

File: src/dmd/backend/cgelem.d [hunk 12]
```diff
@@ -3821,6 +3828,177 @@ private elem * elnegass(elem *e, goal_t goal)
...
+        }
+        e = e.EV.E2;
+    }
+}
+
+/***************************
+ * Count number of OPstrctor in the expression.
+ */
+
+size_t el_countstrctor(elem *e)
```
Comment (WalterBright) [inline src/dmd/backend/cgelem.d]: private?

**Comment by WalterBright** [comment]
I'm uncomfortable with changing the back end like this. For one thing, it's effect on the optimizer is unknown. By doing it in the glue layer, these concerns would vanish. I'm wondering if the code added to cgelem.d could be done in e2ir.d.

**Comment by SSoulaimane** [comment]
@WalterBright it was literally doing the exact same thing in `e2ir.d` inside `elAssign()`, it just didn't blend there naturally. It's like moving something which naturally fits in `cgelem.d` to `e2ir`, it will work for sure but it would raise questions why is tree shifting in `e2ir.d`. These additions only affect the `OPstrctor` and `OPstrthis` nodes which were only used in DMC. This is as clean as I can think of right now. Other than this I can do some manual work in a dozen locations, it wouldn't be as clean though.

**Comment by SSoulaimane** [comment]
@WalterBright I mcved it back to `e2ir.d`, the code added itself copies side effect to a temporary without calling the copy constructor. The good news is nowhere in the codegen structs are being copied copied to temporaries, it only happens in `e2ir.d` `s2ir.d` which makes the frontend solution that @ibuclaw mentioned a plausible solution.

**Comment by ibuclaw** [comment]
> @ibuclaw mentioned a plausible solution.

Ultimately I'm not a fan of having the codegen second guessing what the frontend really means when it is compiling a specific pattern.

We already have various forms of AssignExp - ConstructExp, BlitExp, etc... We could always add a new ast node, i.e CopyConstructExp that replaces the current frontend code:
```
Container __copytmp5 = __copytmp5 = 0 , __copytmp5.this(v);) , __copytmp5
```
This will keep the implementation ultimately in the glue/code generation, however it will remove the ast pattern matching and possibly need for IR rewriting to be done later on.

**Comment by SSoulaimane** [comment]
That's a small part of the problem. The pattern itself which is generated by `doCopyOrMove()` doesn't apply to temporaries created in the glue since there is no semantic there.

**Comment by ibuclaw** [comment]
Only in the case of return statements though?

Correct me if I remember wrong, in any step here.

In the most simplistic case.
```
a = b; // AssignExp(a, b)
auto a = b; // ConstructExp(a, b)
```
Therefore one can express
```
a = c; // CopyConstructExp(a, c, __cpctor)
```
This also works fine if `a = c` is an array type, as the codegen can generate a call to the cpctor on each element in `a`.

In the case of return expressions
```
return c; // ReturnStatement(ConstructExp(tmp, c, __cpctor))
```
If there's a way to mark temporary var decls as being the expression of a return statement `(STCtemp | STCreturnvalue)` then we can set-up tmp as just being an alias to the return slot `<retval>`.  Being a decl with a cpctor, the type is marked as addressable, which tells the backend to not create temporaries/elide copies or else ICE (because we generated code wrong/in a way that couldn't avoid a temporary).

Leaving the complex case for last, objects with cpctor/postblit fields.

There's already recognition of:
```
a = A(1); // ConstructExp(a, A().__ctor(1))
```
Which sets `a` as the target of `A()`, and so gets written directly to the variable instead of a temporary, and subsequently calls `a.__ctor(1)`.

Going from there to the following:
```
a = A(c, 1); // ConstructExp(a, A(c, 1))
```
Is not that big a step, albeit the difference is in the above you must expand the constructor to individual direct assignments.
```
a = 0; // memset()
a.f1.__cpctor(c);
a.f2 = 1;
```
If you are going this far, then might as well have this expansion in the frontend.

All that I mention above though goes on the assumption that the frontend won't insert magical __tmp vars, which only serves to over complicate the matter.

**Comment by SSoulaimane** [comment]
The selling point of the current way `doCopyOrMove()` does it, is the frontend takes care of selecting whether there is a copy constructor or a postblit and checks whether to actually call it or not, but it's not much work really it can fit in a small function in the glue. If the expression is a function call which returns on the stack you use the hidden pointer directly no need to call the copy contructor, likewise with struct literals, for the rest there is default initialization followed by copy construction.

> Only in the case of return statements though?

Everywhere you find `doCopyOrMove()` being called the destination is unknown to the frontend, this includes function arguments, array and struct literals, concatenation expressions, and the hidden parameter for function returns.

Your solution simply sets a function to be called lets the glue do the logic.

**Comment by SSoulaimane** [comment]
@ibuclaw semaphore failure is cause by this GDC bug https://issues.dlang.org/show_bug.cgi?id=20342.

**Comment by ibuclaw** [comment]
The codegen is still better placed to know when a temporary value is needed or not. The current behaviour of the frontend generating ast that creates a temporary only for the codegen to say actually you didn't mean that and then proceed to disentangle the mess that it has consumed is not a desirable solution.

As far as I can tell, the only reason that the frontend creates the whole `(tmp = 0, tmp.__cpctor(a), tmp` or `(tmp = a, tmp.__postblit(), tmp)` is for want of a better way to represent the action of constructing without copying. Both could be simplified into one ast node. Then the glue will provide the ultimate target of the expression, i.e
```
memset(&<retval>, 0, 16);
__cptor(&<retval>, a);
return <retval>;
```

**Comment by SSoulaimane** [comment]
@ibuclaw This approach works. But in general I can see the benefit of moving things to the fronted or to a pre glue phase which is shared between compilers. This would allow GDC and LDC to catchup much quicker with DMD and focus more about the specificities of each backend rather than repeat the repetitive stuff.

**Comment by RazvanN7** [comment]
This seems to have been abandoned. The issue is referencing this PR, therefore anyone that picks it up, feel free to reopen.

**Pull request closed without merging**